NUMERIC  column doesn't resolve to BigDecimal using SUM or Binary Expression

### SQLDelight Version

2.0.2

### SQLDelight Dialect

PostgreSql

### Describe the Bug

a) When a column is declared NUMERIC and the SUM function is used, the resolver chooses Kotlin.Long as the type
https://github.com/cashapp/sqldelight/blob/0d41d969506b29f59ec6abc5eed9d09e963cbb85/dialects/postgresql/src/main/kotlin/app/cash/sqldelight/dialects/postgresql/PostgreSqlTypeResolver.kt#L134-L141

Missing NUMERIC mapping as defaults to Long

b) Binary Expressions
``` sql
 CREATE TABLE Test (
      a NUMERIC,
      b NUMERIC);

SELECT a + b FROM Test;
```

Missing NUMERIC mapping from https://github.com/cashapp/sqldelight/blob/0d41d969506b29f59ec6abc5eed9d09e963cbb85/dialects/postgresql/src/main/kotlin/app/cash/sqldelight/dialects/postgresql/PostgreSqlTypeResolver.kt#L251

### Stacktrace

b) `Caused by: java.util.NoSuchElementException: Array contains no element matching the predicate at app.cash.sqldelight.dialect.api.TypeResolverKt.encapsulatingType(TypeResolver.kt:115)`

--- a/dialects/postgresql/src/main/kotlin/app/cash/sqldelight/dialects/postgresql/PostgreSqlTypeResolver.kt
+++ b/dialects/postgresql/src/main/kotlin/app/cash/sqldelight/dialects/postgresql/PostgreSqlTypeResolver.kt
@@ -99,6 +99,7 @@ class PostgreSqlTypeResolver(private val parentResolver: TypeResolver) : TypeRes
       INTEGER,
       BIG_INT,
       REAL,
+      PostgreSqlType.NUMERIC,
       TEXT,
       BLOB,
       TIMESTAMP_TIMEZONE,
@@ -113,6 +114,7 @@ class PostgreSqlTypeResolver(private val parentResolver: TypeResolver) : TypeRes
       PostgreSqlType.INTEGER,
       BIG_INT,
       REAL,
+      PostgreSqlType.NUMERIC,
       TIMESTAMP_TIMEZONE,
       TIMESTAMP,
     )
@@ -124,19 +126,19 @@ class PostgreSqlTypeResolver(private val parentResolver: TypeResolver) : TypeRes
       if (isArrayType(exprType)) {
         exprType
       } else {
-        encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, TEXT, BLOB, nullability = { exprListNullability ->
+        encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, PostgreSqlType.NUMERIC, TEXT, BLOB, nullability = { exprListNullability ->
           exprListNullability.all { it }
         })
       }
     }
-    "max" -> encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, TEXT, BLOB, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
-    "min" -> encapsulatingTypePreferringKotlin(exprList, BLOB, TEXT, SMALL_INT, INTEGER, PostgreSqlType.INTEGER, BIG_INT, REAL, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
+    "max" -> encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, PostgreSqlType.NUMERIC, TEXT, BLOB, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
+    "min" -> encapsulatingTypePreferringKotlin(exprList, BLOB, TEXT, SMALL_INT, INTEGER, PostgreSqlType.INTEGER, BIG_INT, REAL, PostgreSqlType.NUMERIC, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
     "sum" -> {
       val type = resolvedType(exprList.single())
-      if (type.dialectType == REAL) {
-        IntermediateType(REAL).asNullable()
-      } else {
-        IntermediateType(INTEGER).asNullable()
+      when (type.dialectType) {
+        REAL -> IntermediateType(REAL).asNullable()
+        PostgreSqlType.NUMERIC -> IntermediateType(PostgreSqlType.NUMERIC).asNullable()
+        else -> IntermediateType(INTEGER).asNullable()
       }
     }
     "to_hex", "quote_literal", "quote_ident", "md5" -> IntermediateType(TEXT)
@@ -176,14 +178,14 @@ class PostgreSqlTypeResolver(private val parentResolver: TypeResolver) : TypeRes
     "json_build_object", "jsonb_build_object",
     -> IntermediateType(TEXT)
     "array_agg" -> {
-      val typeForAgg = encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, TEXT, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
+      val typeForAgg = encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, PostgreSqlType.NUMERIC, TEXT, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
       arrayIntermediateType(typeForAgg)
     }
     "string_agg" -> IntermediateType(TEXT)
     "json_array_length", "jsonb_array_length" -> IntermediateType(INTEGER)
     "jsonb_path_exists", "jsonb_path_match", "jsonb_path_exists_tz", "jsonb_path_match_tz" -> IntermediateType(BOOLEAN)
     "currval", "lastval", "nextval", "setval" -> IntermediateType(BIG_INT)
-    "generate_series" -> encapsulatingType(exprList, INTEGER, BIG_INT, REAL, TIMESTAMP_TIMEZONE, TIMESTAMP)
+    "generate_series" -> encapsulatingType(exprList, INTEGER, BIG_INT, REAL, PostgreSqlType.NUMERIC, TIMESTAMP_TIMEZONE, TIMESTAMP)
     "regexp_count", "regexp_instr" -> IntermediateType(INTEGER)
     "regexp_like" -> IntermediateType(BOOLEAN)
     "regexp_replace", "regexp_substr" -> IntermediateType(TEXT)
@@ -193,7 +195,7 @@ class PostgreSqlTypeResolver(private val parentResolver: TypeResolver) : TypeRes
     "websearch_to_tsquery" -> IntermediateType(TEXT)
     "rank", "dense_rank", "row_number" -> IntermediateType(INTEGER)
     "ntile" -> IntermediateType(INTEGER).asNullable()
-    "lag", "lead", "first_value", "last_value", "nth_value" -> encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, TEXT, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
+    "lag", "lead", "first_value", "last_value", "nth_value" -> encapsulatingTypePreferringKotlin(exprList, SMALL_INT, PostgreSqlType.INTEGER, INTEGER, BIG_INT, REAL, PostgreSqlType.NUMERIC, TEXT, TIMESTAMP_TIMEZONE, TIMESTAMP, DATE).asNullable()
     else -> null
   }
 
@@ -265,6 +267,7 @@ class PostgreSqlTypeResolver(private val parentResolver: TypeResolver) : TypeRes
           INTEGER,
           BIG_INT,
           REAL,
+          PostgreSqlType.NUMERIC,
           TEXT,
           BLOB,
           BOOLEAN,

--- a/sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/InterfaceGeneration.kt
+++ b/sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/InterfaceGeneration.kt
@@ -1147,6 +1147,7 @@ class InterfaceGeneration {
     |inr INTEGER,
     |jsn JSON,
     |jsb JSON,
+    |num NUMERIC,
     |tim TIME,
     |tms TIMESTAMP,
     |tmz TIMESTAMPTZ,
@@ -1167,6 +1168,7 @@ class InterfaceGeneration {
     |inr IS NOT NULL AS has_integer,
     |jsn IS NOT NULL AS has_json,
     |jsb IS NOT NULL AS has_jsob,
+    |num IS NOT NULL AS has_num,
     |sml IS NOT NULL AS has_smallint,
     |tim IS NOT NULL AS has_time,
     |tms IS NOT NULL AS has_timestamp,
@@ -1205,6 +1207,7 @@ class InterfaceGeneration {
     |    has_integer: Boolean,
     |    has_json: Boolean,
     |    has_jsob: Boolean,
+    |    has_num: Boolean,
     |    has_smallint: Boolean,
     |    has_time: Boolean,
     |    has_timestamp: Boolean,
@@ -1222,6 +1225,7 @@ class InterfaceGeneration {
     |  |inr IS NOT NULL AS has_integer,
     |  |jsn IS NOT NULL AS has_json,
     |  |jsb IS NOT NULL AS has_jsob,
+    |  |num IS NOT NULL AS has_num,
     |  |sml IS NOT NULL AS has_smallint,
     |  |tim IS NOT NULL AS has_time,
     |  |tms IS NOT NULL AS has_timestamp,
@@ -1247,11 +1251,12 @@ class InterfaceGeneration {
     |      cursor.getBoolean(11)!!,
     |      cursor.getBoolean(12)!!,
     |      cursor.getBoolean(13)!!
+    |      cursor.getBoolean(14)!!
     |    )
     |  }
     |
     |  public fun selectIsNotNull(): Query<SelectIsNotNull> = selectIsNotNull { has_bigint, has_boolean,
-    |      has_byte, has_date, has_integer, has_json, has_jsob, has_smallint, has_time, has_timestamp,
+    |      has_byte, has_date, has_integer, has_json, has_jsob, has_num, has_smallint, has_time, has_timestamp,
     |      has_timestamptz, has_tsvector, has_uuid, has_varchar ->
     |    SelectIsNotNull(
     |      has_bigint,
@@ -1261,6 +1266,7 @@ class InterfaceGeneration {
     |      has_integer,
     |      has_json,
     |      has_jsob,
+    |      has_num,
     |      has_smallint,
     |      has_time,
     |      has_timestamp,
@@ -1272,6 +1278,118 @@ class InterfaceGeneration {
     |  }
     |}
     |
+      """.trimMargin(),
+    )
+  }
+
+  @Test
+  fun `postgres using numeric returns BigDecimal`() {
+    val file = FixtureCompiler.parseSql(
+      """
+        |CREATE TABLE sales (
+        |  product_id INTEGER,
+        |  sale_date DATE,
+        |  sale_amount NUMERIC
+        |);
+        |
+        |select:
+        |SELECT
+        |  sum(sale_amount) AS sum_amount,
+        |  sale_amount * sale_amount AS product_amount,
+        |  array_agg(sale_amount) AS agg_amount,
+        |  generate_series(sale_amount, 2) AS generate_amount
+        |FROM sales
+        |GROUP BY sale_amount;
+      """.trimMargin(),
+      temporaryFolder,
+      fileName = "NumericFunctions.sq",
+      dialect = PostgreSqlDialect(),
+    )
+
+    val table = file.tables(false).single()
+    val tableGenerator = TableInterfaceGenerator(table)
+
+    assertThat(tableGenerator.kotlinImplementationSpec().toString()).isEqualTo(
+      """
+      |public data class Sales(
+      |  public val product_id: kotlin.Int?,
+      |  public val sale_date: java.time.LocalDate?,
+      |  public val sale_amount: java.math.BigDecimal?,
+      |)
+      |
+      """.trimMargin(),
+    )
+
+    val query = file.namedQueries.first()
+    val queryGenerator = QueryInterfaceGenerator(query)
+    assertThat(queryGenerator.kotlinImplementationSpec().toString()).isEqualTo(
+      """
+      |public data class Select(
+      |  public val sum_amount: java.math.BigDecimal?,
+      |  public val product_amount: java.math.BigDecimal?,
+      |  public val agg_amount: kotlin.Array<java.math.BigDecimal?>,
+      |  public val generate_amount: java.math.BigDecimal,
+      |)
+      |
+      """.trimMargin(),
+    )
+  }
+
+  @Test
+  fun `postgres using numeric window function returns BigDecimal`() {
+    val file = FixtureCompiler.parseSql(
+      """
+        |CREATE TABLE sales (
+        |  product_id INTEGER,
+        |  sale_date DATE,
+        |  sale_amount NUMERIC
+        |);
+        |
+        |select:
+        |SELECT
+        |  product_id,
+        |  sale_date,
+        |  lag(sale_amount, 1) OVER (PARTITION BY product_id ORDER BY sale_date) AS prev_sale_amount,
+        |  lead(sale_amount, 1) OVER (PARTITION BY product_id ORDER BY sale_date) AS next_sale_amount,
+        |  first_value(sale_amount) OVER (PARTITION BY product_id ORDER BY sale_date) AS first_sale_amount,
+        |  last_value(sale_amount) OVER (PARTITION BY product_id ORDER BY sale_date RANGE BETWEEN UNBOUNDED PRECEDING AND UNBOUNDED FOLLOWING) AS last_sale_amount,
+        |  nth_value(sale_amount, 2) OVER (PARTITION BY product_id ORDER BY sale_date) AS second_sale_amount
+        |FROM sales
+        |ORDER BY product_id, sale_date;
+      """.trimMargin(),
+      temporaryFolder,
+      fileName = "NumericWindowsFunctions.sq",
+      dialect = PostgreSqlDialect(),
+    )
+
+    val table = file.tables(false).single()
+    val tableGenerator = TableInterfaceGenerator(table)
+
+    assertThat(tableGenerator.kotlinImplementationSpec().toString()).isEqualTo(
+      """
+      |public data class Sales(
+      |  public val product_id: kotlin.Int?,
+      |  public val sale_date: java.time.LocalDate?,
+      |  public val sale_amount: java.math.BigDecimal?,
+      |)
+      |
+      """.trimMargin(),
+    )
+
+    val query = file.namedQueries.first()
+    val queryGenerator = QueryInterfaceGenerator(query)
+    assertThat(queryGenerator.kotlinImplementationSpec().toString()).isEqualTo(
+      """
+      |public data class Select(
+      |  public val product_id: kotlin.Int?,
+      |  public val sale_date: java.time.LocalDate?,
+      |  public val prev_sale_amount: java.math.BigDecimal?,
+      |  public val next_sale_amount: java.math.BigDecimal?,
+      |  public val first_sale_amount: java.math.BigDecimal?,
+      |  public val last_sale_amount: java.math.BigDecimal?,
+      |  public val second_sale_amount: java.math.BigDecimal?,
+      |)
+      |
       """.trimMargin(),
     )
   }

--- a/sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/InterfaceGeneration.kt
+++ b/sqldelight-compiler/src/test/kotlin/app/cash/sqldelight/core/queries/InterfaceGeneration.kt
@@ -1250,14 +1250,14 @@ class InterfaceGeneration {
     |      cursor.getBoolean(10)!!,
     |      cursor.getBoolean(11)!!,
     |      cursor.getBoolean(12)!!,
-    |      cursor.getBoolean(13)!!
+    |      cursor.getBoolean(13)!!,
     |      cursor.getBoolean(14)!!
     |    )
     |  }
     |
     |  public fun selectIsNotNull(): Query<SelectIsNotNull> = selectIsNotNull { has_bigint, has_boolean,
-    |      has_byte, has_date, has_integer, has_json, has_jsob, has_num, has_smallint, has_time, has_timestamp,
-    |      has_timestamptz, has_tsvector, has_uuid, has_varchar ->
+    |      has_byte, has_date, has_integer, has_json, has_jsob, has_num, has_smallint, has_time,
+    |      has_timestamp, has_timestamptz, has_tsvector, has_uuid, has_varchar ->
     |    SelectIsNotNull(
     |      has_bigint,
     |      has_boolean,

